src/backend/utils/misc/queryenvironment.c

```diff
@@ -241,6 +246,83 @@ bool has_existing_enr_relations()
 	return false;
 }
 
+/* This only stores catcache inval messages meant for ENR tables in the current transaction. */
+void SaveCatcacheMessage(int cacheId,
+							 uint32 hashValue,
+							 Oid dbId)
+{
+	SharedInvalCatcacheMsg *msg = (SharedInvalCatcacheMsg *) palloc0(sizeof(SharedInvalCatcacheMsg));
```

Comment:
what if this list grows very large? also can there be case where hash value is not unique across entries?

Reply:
1. The list growing large would be no different than the size of the inval message queue growing very large. I'm not aware of any cap on number of invalidation messages in one transaction.
2. We can't alter the hash value here because it must match the hash that is expected by the existing SI code in `SIInsertDataEntries`. So if there is a collision in PG, then there would be one here.

Reply:
PG inval queue is not used for lookup. Here, we do lookup whenever a new entry is added. So, we need to think if this can be an issue in average case as we are doing sequential loopkup vs using hash map etc.

Reply:
i think it is not an issue. the number of catalog entries that can be in one transaction should basically be limited by the number of temp tables created in that transaction. 

if we were to switch to a hash table, we then would need to take the overhead of initializing and emptying the hash table for each transaction, which could get expensive, even for small transactions with only one temp table create.